sklearn/utils/_array_api.py

```diff
@@ -422,47 +505,109 @@ def _add_to_diagonal(array, value, xp):
-        else:
-            sample_weight_np = None
-        return float(numpy.average(sample_score_np, weights=sample_weight_np))
+        xp, _ = get_namespace(*input_arrays)
 
-    if not xp.isdtype(sample_score.dtype, "real floating"):
-        # We move to cpu device ahead of time since certain devices may not support
-        # float64, but we want the same precision for all devices and namespaces.
-        sample_score = xp.astype(xp.asarray(sample_score, device="cpu"), xp.float64)
+    device_ = device(*input_arrays)
 
-    if sample_weight is not None:
-        sample_weight = xp.asarray(
-            sample_weight, dtype=sample_score.dtype, device=device(sample_score)
-        )
-        if not xp.isdtype(sample_weight.dtype, "real floating"):
-            sample_weight = xp.astype(sample_weight, xp.float64)
+    if _is_numpy_namespace(xp) and normalize:
+        return xp.asarray(numpy.average(a, axis=axis, weights=weights))
 
-    if normalize:
-        if sample_weight is not None:
-            scale = xp.sum(sample_weight)
-        else:
-            scale = sample_score.shape[0]
-        if scale != 0:
-            sample_score = sample_score / scale
+    a = xp.asarray(a, device=device_)
+    if weights is not None:
+        weights = xp.asarray(weights, device=device_)
+
+    output_dtype = None
+    output_dtype_name = None
 
-    if sample_weight is not None:
-        return float(sample_score @ sample_weight)
+    if xp.isdtype(a.dtype, "bool"):
+        a = xp.astype(a, xp.int32)
+    if weights is not None and xp.isdtype(weights.dtype, "bool"):
+        weights = xp.astype(weights, xp.int32)
+
+    if any(
+        (
+            (input_array is not None)
+            and (
+                (not xp.isdtype(input_array.dtype, "numeric"))
+                or xp.isdtype(input_array.dtype, "complex floating")
+            )
+        )
+        for input_array in [a, weights]
+    ):
+        raise ValueError("Expecting only boolean, integral or real floating values.")
+
+    if weights is None and xp.isdtype(a.dtype, "integral"):
+        output_dtype_name = "float64"
+    elif weights is None:
+        output_dtype = a.dtype
+    elif xp.isdtype(a.dtype, "real floating") and xp.isdtype(
+        weights.dtype, "real floating"
+    ):
+        output_dtype = (
+            a.dtype
+            if (xp.finfo(a.dtype).bits >= xp.finfo(a.dtype).bits)
+            else weights.dtype
+        )
     else:
-        return float(xp.sum(sample_score))
+        output_dtype_name = "float64"
+
+    cast_to_float64 = (output_dtype_name == "float64") or (
+        xp.finfo(output_dtype).bits == 64
+    )
+
+    if cast_to_float64 and not _supports_dtype(xp, device_, "float64"):
+        a = _convert_to_numpy(a, xp)
+        if weights is not None:
+            weights = _convert_to_numpy(weights, xp)
+        return xp.asarray(
+            _average(a, axis=axis, normalize=normalize, weights=weights),
+            dtype=xp.float32,
```

Comment:
Good question. `numpy.average` stays in `np.float16` so I think we might want to replicate the same behavior. However it's quite likely that using `numpy.average` on `np.float16` data will result in non numerically stable results.

Reply:
If we don't want to decide now, we can just raise `NotImplementedError` in this case and delay the decision to the day where we actually need this in scikit-learn. It's quite possible that this will never happen.

Reply:
This particular branch of code is only ran `if cast_to_float64` (seeing the _if_ condition that encloses it), but the device does not support `float64` so it falls back to `float32`, because it is the highest available precision right after `float64`, the actual precision of all of the inputs is not relevant once this path is triggered. 

If `cast_to_float64` code path is not triggered, the actual type of the inputs will indeed be used to determine the type of the output (e.g all float16 inputs should give float16 output).

The casting rules are exhaustively described in [numpy.average documentation page](https://numpy.org/doc/stable/reference/generated/numpy.average.html).